### ISSUE TITLE
Refactor a HotReloader out of BlueprintGenericEditor

### ISSUE THREAD

**Issue by nick-thompson** [nick-thompson/blueprint#27]
The `BlueprintGenericEditor` implements hot reloading by polling the file system to see if the js bundle has changed. It's super simple, and there's no reason that it has to be tied to a notion of a plugin editor (it should be usable no matter what type of app you're using). So I think just pulling that out into a separate class which can be composed wherever you need it makes sense.

```
HotReloader(std::unique_ptr<ReactApplicationRoot>& appRoot, const juce::File& bundle)
```

Just run the timer, check the bundle, and `reset()` the `appRoot` there. Then BlueprintGenericEditor just composes the HotReloader for this functionality

**Comment by JoshMarler** [comment]
Hey @nick-thompson, This framework is so flipping good. I'm actually working on this as part of a project I'm working on at my current shop. If you're still after this I can drop over a few pull request with some of my changes when I'm satisfied things are working well. 

I also have a few GCC fixes as I've been compiling on Linux

**Comment by JoshMarler** [comment]
As an aside. Presumably we need to notify clients of HotReloader at two points. Going off of the example code in BlueprintGenericEditor::assignNewAppRoot:

1. When resetting the appRoot unique_ptr so that clients can then call registerNativeMethod etc.
2. Once appRoot->evaluate(bundleFile) has been called so clients can addAndMakeVisible etc.

i.e. Something along the lines of the below:
`    void HotReloader::timerCallback()
    {
        const auto lmt = bundleFile.getLastModificationTime();

        if (lmt > bundleLastModifiedTime)
        {
            // Sanity check
            jassert(bundleFile.existsAsFile());

            // Assign a fresh appRoot
            appRoot = std::make_unique<ReactApplicationRoot>();
            listeners.call([=](Listener &l) { l.ReloadingAppRootBundle(appRoot.get(), bundleFile); });

            appRoot->evaluate(bundleFile.getFullPathName());
            listeners.call([=](Listener &l) { l.AppRootBundleReloaded(appRoot.get(), bundleFile); });

            // At this point the appRoot may have been removed due to an error evaluating
            // the bundle, so we check for that case and halt if necessary
            if (!appRoot)
                return;

            bundleLastModifiedTime = lmt;
        }
    }`

Where the pointer passed to AppRootBundleReloaded is potentially nullptr in the event of an error (Specified in the Documentation/Contract of the class)

I'm not entirely clear on the expected usage of this class yet. I've used the above pattern in my initial experiments with a ReactBasedComponent class that includes a HotLoader via composition. The ReactBasedComponent take out the AudioProcessorEditor side of BlueprintGenericEditor so that clients can simply call the below and ignore the boiler plate of bundle loading, hot reload and error handling:

`ReactBasedComponent myComponent(bundleFile);`

Still woking out how best to deal with registering native methods and dispatching events etc. etc. 

Cheers

**Comment by nick-thompson** [comment]
Hey @JoshMarler thanks so much, I would very happily welcome your changes!

Totally hear you, I haven't sorted out the best API for this class yet either. I like the callbacks you have there for sure, and I also like the decoupling from `BlueprintGenericEditor`. It sounds like `BlueprintGenericEditor` could just simply consume `ReactBasedComponent`? And maybe the hooks for native methods/event dispatch are just some lambda assignments:

```
mReactComponent = std::make_unique<ReactBasedComponent>(bundleFile);
mReactComponent->beforeBundleEval([this]() {
  // set naative methods
});
mReactComponent->afterBundleEval([this]() {
  // dispatch events
});
```

What do you think? Also, realizing just now as I'm typing, do you think it would be effective to just roll this stuff into the `ReactApplicationRoot` class?

**Comment by JoshMarler** [comment]
Hey Nick!

Yeah I toyed between lambda and listener patterns, std::function member sounds fine to me.

In regards to rolling the changes into ReactApplicationRoot, certainly seems worth a go. I'll write up both approaches (ReactBasedComponent and ReactApplicationRoot) and see which looks nicer in the context of this project I'm working on. Then maybe I can drop you an MR for each and you can critique and pick the best approach?

Most likely I'll drop it over tomorrow evening. 

I'll send separate patches over for the GCC fixes, FYI, one required a Yoga submodule update. Enums and bitfields caused GCC headaches.

**Comment by nick-thompson** [comment]
That sounds great, thanks so much for your help

**Comment by JoshMarler** [comment]
Ok so this might land more like a couple days from now. Bugs to kill at work :-(

**Comment by nick-thompson** [comment]
No problem, totally understand :)

**Comment by JoshMarler** [comment]
Just thinking about finishing this off @nick-thompson , I wonder whether it actually makes sense to have a HotReloader class at all? I'm unlikely to ever us it outside of ReactApplicationRoot and HotReloader would require callbacks on bundle change to signal ReactApplicationRoot anyway. 

Perhaps it makes more sense for ReactApplicationRoot to do the"hot reloading" and simply have an enableHotReload()? The constructor could even call enableHotReload() if JUCE_DEBUG is defined?

**Comment by JoshMarler** [comment]
Might be easier if I just drop a pull request and you can pull it apart

**Comment by nick-thompson** [comment]
Yea, I think you're right– rolling it into `ReactApplicationRoot` makes the most sense. And I think with the before and after callback hooks it will be as flexible as anybody might need.



> Perhaps it makes more sense for ReactApplicationRoot to do the"hot reloading" and simply have an enableHotReload()? The constructor could even call enableHotReload() if JUCE_DEBUG is defined?

:+1: Love that

> If this approach was taken would it make sense to just pass the bundle to ReatApplicationRoot's constructor rather than having evaluate() public. Are there use cases for calling evaluate multiple times?

I think there should always be a way to reach `ReactApplicationRoot`'s `EcmascriptEngine` just so you can always interact with the js closely, and you could call `evaluate` directly there. The `evaluate` on ReactApplicationRoot was just for convenience, but with the `EcmascriptEngine` a public member it's really not that much more convenient :) We could remove it. I think there are cases for multiple evals, but none that I'm actively using. For example, I could imagine calling `evaluate` right away at start up to polyfill the js environment with things like `fetch` (maybe even in the EcmascriptEngine constructor), and then calling a second time to run your own bundle.

I also want to call out that if we roll hot reloading into ReactApplicationRoot, then pulling out a ViewManager might be a very helpful first step: https://github.com/nick-thompson/blueprint/issues/9 and https://github.com/nick-thompson/blueprint/blob/master/blueprint/core/blueprint_ReactApplicationRoot.h#L67. For example, if the `ReactApplicationRoot` owns a `std::unique_ptr<ViewManager>` member, then during the hot reloading "reset" we could just replace that member with a fresh `ViewManager` instance and that would probably be all of the cleanup we need. It's not a trivial refactor to get that ViewManager into its own thing but it could make this really nice

**Comment by JoshMarler** [comment]
OK Awesome. Food for throught.

> For example, I could imagine calling evaluate right away at start up to polyfill the js environment with things like fetch (maybe even in the EcmascriptEngine constructor), and then calling a second time to run your own bundle.

This is a super useful thing to know actually. It could open up lazy load opportunities with things like reusable component modules in a project with multiple bundles.

**Comment by JoshMarler** [comment]
Hey Nick. Just an FYI.

I'm having a crack at pulling out the ViewManager as we speak. I hit a point in the HotReload refactor where it's actually necessary and having it as a unique_ptr as you've suggested will indeed make hot reloading in ReactApplicationRoot much more straight forward. 

I'll drop a pull request in as soon as I'm done as maybe we can pull it apart together. I have a couple interesting issues to discuss around hot reload behaviour already.

**Comment by JoshMarler** [comment]
Would you be up for an MR with a couple of commits? The first will be the ViewManager separation and then later commits will be hot reloading managed by ReactApplicationRoot? It's useful for me to think about both and how the will work together and I think it would be good to discuss the two changes as a pair. Hopefully the diff won't be too huge. 

ViewManager is actually coming out quite nicely. The hot reload complicates slightly but should all be solveable.

**Comment by nick-thompson** [comment]
🙌 awesome! Yea, feel free to make multiple commits and even multiple merge requests. I'm definitely in favor of breaking up the work along lines that are sensible to reason about

**Comment by JoshMarler** [comment]
One interesting thing from a client perspective is whether it's nicer to have ViewManager as a unique_ptr member or as a class inherited by ReactApplicationRoot. The stuff only used by ReactApplicationRoot can become protected and we could just make registerViewType and the like public?

It's that or a ReactApplicationRoot::registerViewType which delegates to the same names function on the ViewManager unique_ptr instance. 

If ReactApplicationRoot inherits ViewManager would need some sort of dispose() or clear() function to facilitate the hot reloads. 

i.e. in a ReactApplicationRoot onBundleChange() callback or similar ReactApplicationRoot would call clear(), do whatever work was required and then call registerNativeRenderingHooks and installNativeViewTypes. 

Happy to send over both approachs in MR

**Comment by nick-thompson** [comment]
I think I'm generally a fan of composition over inheritance, but especially here I think it makes sense to lean on composition because then dispose() or clear() just becomes `viewManager.reset(new ViewManager())` (and I guess resetting the ecmascript engine instance).

I think `registerViewType` is the only public method on ReactApplicationRoot that would be basically a proxy to view manager, so it seems like the consequences of choosing composition here would be pretty small

**Comment by JoshMarler** [comment]
Cool I'm down with that. Will finish off with that approach.

**Comment by JoshMarler** [comment]
Alrighty. Done with the ViewManager pretty much. Will throw over PR shortly. Has some TODO comments I thought we could discuss and I'll remove them and push up resolutions once discussed. Just running some tests/sanity checks

**Comment by nick-thompson** [comment]
Closing this, resolved in #54 🎉
